### PR TITLE
tchannel golang initial implementation

### PR DESCRIPTION
**Pull request by mmihic**
Initial golang tchannel implementation.  See [README](golang/README.md) for high level description, and  [TODO](golang/TODO.md) for list of TODO items.

### CODE REVIEW

File: golang/README.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+NOTE: `master:golang` is **not yet stable**
+
+## Example
+
+```golang
+import (
+    "encoding/json"
+    "time"
+    "github.com/uber/tchannel"
```
Comment (abhinav) [inline golang/README.md]: Shouldn't this be tchannel/golang?

File: golang/README.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    // Make a call to ourselves, with a timeout of 10s
+    ctx, cancel := context.WithTimeout(context.Background(), time.Second * 10)
+    defer cancel()
+
+    var responseHeaders Headers
+    var pong Pong
+    if err := client.RoundTrip("localhost:8050", "PingService", "ping",
+        NewJSONOutput(Headers{}),  NewJSONOutput(Ping{Message: "Hello world"}),
+        NewJSONInput(&responseHeaders), NewJSONInput(&pong)); err != nil {
```
Comment (abhinav) [inline golang/README.md]: These calls probably need to be prepended with `tchannel.`

File: golang/Godeps/Godeps.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
+	"ImportPath": "code.uber.internal/personal/mmihic/tchannel-go",
```
Comment (sectioneight) [inline golang/Godeps/Godeps.json]: We probably want to scrub this

Comment (mmihic) [inline golang/Godeps/Godeps.json]: oops, yes, need to cleanup all of these imports.

File: golang/Makefile [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	mkdir -p $(BUILD)
+	go test $(PKGS) $(TEST_ARG) -parallel=4
+
+# target: cover - Runs tests under code coverage
+cover: clean
+	echo Testing packages:
+	mkdir -p $(BUILD)
+	go test ./ $(TEST_ARG)  -coverprofile=$(BUILD)/coverage.out
+	go tool cover -html=$(BUILD)/coverage.out
+	# TODO(mmihic): Temporarily disabled while working out false positives
```
Comment (sectioneight) [inline golang/Makefile]: is this still a TODO?

File: golang/arguments.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package tchannel
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+)
+
+// An Input is able to read an argument from a call body
```
Comment (sectioneight) [inline golang/arguments.go]: Suggest making comments golint-compatible (e.g. start the comment with the type name you're describing):

```
// Input describes a type...
```

File: golang/channel.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+// A TChannel is a bi-directional connection to the peering and routing network.  Applications
+// can use a TChannel to make service calls to remote peers via BeginCall, or to listen for incoming calls
+// from peers.  Once the channel is created, applications should call the ListenAndHandle method to
+// listen for incoming peer connections.  Because channels are bi-directional, applications should call
+// ListenAndHandle even if they do not offer any services
+type TChannel struct {
+	log               *logging.Logger
+	hostPort          string
```
Comment (sectioneight) [inline golang/channel.go]: Do we want any of these fields to be exported? e.g. `HostPort`?

File: golang/Makefile (hunk 4 above)
Comment (mmihic) [inline golang/Makefile]: yes annoyingly go-logging results in false positives from vet due to having an Error method.  There's a TODO to remove the go-logging dependency which will resolve; I held off doing that in this branch but can do so.

File: golang/channel.go (hunk 6 above)
Comment (mmihic) [inline golang/channel.go]: Prefer not to export immutable fields.

Comment (mmihic) [inline golang/channel.go]: I can add a method for this if we think it we want access.

File: golang/checksum.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	switch t {
+	case ChecksumTypeNone:
+		return NullChecksum{}
+	case ChecksumTypeCrc32:
+		return NewCrc32Checksum()
+	case ChecksumTypeFarmhash:
+		return nil
+	}
+
+	return nil
```
Comment (sectioneight) [inline golang/checksum.go]: Might make more sense as the `default:` case

File: golang/checksum.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+// Creates a new Checksum of the given type
+func (t ChecksumType) New() Checksum {
+	switch t {
+	case ChecksumTypeNone:
+		return NullChecksum{}
+	case ChecksumTypeCrc32:
+		return NewCrc32Checksum()
+	case ChecksumTypeFarmhash:
+		return nil
```
Comment (sectioneight) [inline golang/checksum.go]: What does this mean for callers? Should there be a TODO about support? Or should we make this a NullChecksum if we're never going to support this?

Comment (mmihic) [inline golang/checksum.go]: There's a TODO for this.

File: golang/connection.go [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+// Information about a TChannel peer
+type PeerInfo struct {
+	// The host and port that can be used to contact the peer, as encoded by net.JoinHostPort
+	HostPort string
+
+	// The logical process name for the peer, used for only for logging / debugging
+	ProcessName string
+}
+
+func (p PeerInfo) String() string {
```
Comment (sectioneight) [inline golang/connection.go]: nice touch

File: golang/arguments.go (hunk 5 above)
Comment (mmihic) [inline golang/arguments.go]: This is already golint-compatible (comments for types can start with an article).  But a bunch of other stuff is not (such as the lack of godoc comments on the ReadFrom/WriteTo implementations).  Will update.

File: golang/TODO.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
+h1. Functionality
+
+[ ] Cancel outbound requests
```
Comment (sectioneight) [inline golang/TODO.md]: This doesn't render properly in github. If you want it to, I think you need to make this a list, e.g. prefix with `- `

File: golang/connection.go [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+		case MessageTypeCallRes:
+			c.outbound.handleCallRes(frame)
+		case MessageTypeCallResContinue:
+			c.outbound.handleCallResContinue(frame)
+		case MessageTypeInitReq:
+			c.handleInitReq(frame)
+		case MessageTypeInitRes:
+			c.handleInitRes(frame)
+		case MessageTypeError:
+			c.handleError(frame)
```
Comment (sectioneight) [inline golang/connection.go]: Looking at the TODO list, I don't see anything about protocol-level ping/pong handling. Am I missing where that is implemented?

File: golang/arguments.go (hunk 5 above)
Comment (sectioneight) [inline golang/arguments.go]: Ah good call. And yea doesn't have to be now, just something we should strive for in the long term.

File: golang/connection.go (hunk 11 above)
Comment (mmihic) [inline golang/connection.go]: Nope, you didn't miss it.  Added to the TODO list.

File: golang/arguments.go (hunk 5 above)
Comment (mmihic) [inline golang/arguments.go]: I'll do it now; I'd rather have the build enforce linting.

**Comment by sectioneight** [comment]
This looks great. I can't do a full review given the total lines of code but I say ship this and then let's follow up on the TODOs. Maybe see if you can get TravisCI running?

File: golang/messages.go [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+	if err != nil {
+		return err
+	}
+
+	m.Headers = callHeaders{}
+	if err := m.Headers.read(r); err != nil {
+		return err
+	}
+
+	serviceNameLen, err := r.ReadUint16()
```
Comment (abhinav) [inline golang/messages.go]: This is being changed to `service~1` so `ReadByte` here instead.

File: golang/messages.go [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+
+	if err := w.WriteByte(m.TraceFlags); err != nil {
+		return err
+	}
+
+	if err := m.Headers.write(w); err != nil {
+		return err
+	}
+
+	if err := w.WriteUint16(uint16(len(m.Service))); err != nil {
```
Comment (abhinav) [inline golang/messages.go]: `WriteByte`

**Comment by Raynos** [comment]
Is this already in master ? can we close this ?

**Pull request closed without merging**